src/NuGet.Core/NuGet.Packaging/Rules/InvalidUndottedFrameworkRule.cs

```diff
@@ -0,0 +1,106 @@
+    {
+        public string MessageFormat { get; }
+
+        public InvalidUndottedFrameworkRule(string messageFormat)
+        {
+            MessageFormat = messageFormat;
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            var set = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+            foreach (var file in builder.GetFiles().Select(t => PathUtility.GetPathWithDirectorySeparator(t)))
+            {
+                set.Add(file);
+            }
+
+            var managedCodeConventions = new ManagedCodeConventions(new RuntimeGraph());
+            var collection = new ContentItemCollection();
+            collection.Load(set.Select(path => path.Replace('\\', '/')).ToArray());
+
+
+            var patterns = managedCodeConventions.Patterns;
+
+            var frameworkPatterns = new List<PatternSet>()
+            {
+                patterns.RuntimeAssemblies,
+                patterns.CompileRefAssemblies,
+                patterns.CompileLibAssemblies,
+                patterns.NativeLibraries,
+                patterns.ResourceAssemblies,
+                patterns.MSBuildFiles,
+                patterns.ContentFiles,
+                patterns.ToolsAssemblies,
+                patterns.EmbedAssemblies,
+                patterns.MSBuildTransitiveFiles
+            };
+            var messages = new HashSet<PackagingLogMessage>();
+
+            foreach (var pattern in frameworkPatterns)
+            {
+                IEnumerable<ContentItemGroup> targetedItemGroups = ContentExtractor.GetContentForPattern(collection, pattern);
+                foreach (ContentItemGroup group in targetedItemGroups)
+                {
+                    foreach (ContentItem item in group.Items)
+                    {
+                        var frameworkString = (string)item.Properties["tfm_raw"];
+                        var framework = (NuGetFramework)item.Properties["tfm"];
+                        try
+                        {
+                            if (framework == null)
+                            {
+                                continue;
+                            }
+
+                            if (framework.Version.Major >= 5 &&
+                                StringComparer.OrdinalIgnoreCase.Equals(FrameworkConstants.FrameworkIdentifiers.NetCoreApp, framework.Framework))
+                            {
+                                var dotIndex = frameworkString.IndexOf('.');
+                                var dashIndex = frameworkString.IndexOf('-');
+                                var frameworkVersionHasDots = (dashIndex > -1 && dotIndex > -1 && dotIndex < dashIndex) || (dashIndex == -1 && dotIndex > -1);
+                                if (!frameworkVersionHasDots)
+                                {
+                                    messages.Add(CreatePackageIssue(item.Path));
```

Comment:
Can we create one warning and just list the offending file paths? 

We have followed `1 vs n` warning approach in the past with: https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Rules/DependenciesGroupsForEachTFMRule.cs. 

Restore also merges the same code warnings on the same target graph (same TFM/RID combo)